File: lib/icb/core/model/odds.ex
defmodule ICB.Core.Model.Odds do
  @moduledoc false

  def list, do: [:small_chance, :unlikely, :fifty_fifty, :likely, :almost_certain]

  @spec resolve(atom(), keyword) :: :no | :yes
  def resolve(odds, opts \\ []) do
    opts = Keyword.merge([rand: &__MODULE__.rand/1], opts)
    rand = Keyword.get(opts, :rand)

    case {odds, rand.(100)} do
      {:small_chance, chance} when chance <= 10 -> :yes
      {:unlikely, chance} when chance <= 25 -> :yes
      {:fifty_fifty, chance} when chance <= 50 -> :yes
      {:likely, chance} when chance <= 75 -> :yes
      {:almost_certain, chance} when chance <= 90 -> :yes
      _ -> :no
    end
  end

  @spec to_string(:almost_certain | :fifty_fifty | :likely | :small_chance | :unlikely) ::
          String.t()
  def to_string(:small_chance), do: "Small Chance"
  def to_string(:unlikely), do: "Unlikely"
  def to_string(:fifty_fifty), do: "50/50"
  def to_string(:likely), do: "Likely"
  def to_string(:almost_certain), do: "Almost Certain"

  @spec parse(String.t()) ::
          :almost_certain | :fifty_fifty | :likely | :small_chance | :unlikely
  def parse("Small Chance"), do: :small_chance
  def parse("Unlikely"), do: :unlikely
  def parse("50/50"), do: :fifty_fifty
  def parse("Likely"), do: :likely
  def parse("Almost Certain"), do: :almost_certain

  @spec rand(pos_integer) :: pos_integer
  def rand(n), do: :rand.uniform(n)
end

File: core/model/odds.go
package model

import "fmt"

/*
Odds 			The answer is ‘yes’ if you roll...
---------------|----------------------------------
Almost Certain	11 or greater
Likely 			26 or greater
50/50 			51 or greater
Unlikely 		76 or greater
Small Chance 	91 or greater
*/

func ResolveOdds(n int, chosenOdds string) NoYes {
	parsedOdds := parseOdds(chosenOdds)
	return resolveOdds(n, parsedOdds)
}

type NoYes bool

func (n NoYes) String() string {
	if n {
		return "Yes"
	}
	return "No"
}

// Odds represents the odds of a roll. It does not participate in calculations.
// It can have only one of the valid values.
type Odds int

func (o Odds) validateOdds() {
	switch o {
	case AlmostCertain:
		return
	case Likely:
		return
	case FiftyFifty:
		return
	case Unlikely:
		return
	case SmallChance:
		return
	}

	panic(fmt.Sprintf("invalid odds: %v", int(o)))
}

func (o Odds) String() string {
	switch o {
	case AlmostCertain:
		return "Almost Certain"
	case Likely:
		return "Likely"
	case FiftyFifty:
		return "50/50"
	case Unlikely:
		return "Unlikely"
	case SmallChance:
		return "Small Chance"
	}

	panic(fmt.Sprintf("invalid odds: %v", int(o)))
}

const (
	AlmostCertain Odds = 11
	Likely        Odds = 26
	FiftyFifty    Odds = 51
	Unlikely      Odds = 76
	SmallChance   Odds = 91
)

func parseOdds(s string) Odds {
	switch s {
	case "Almost Certain":
		return AlmostCertain
	case "Likely":
		return Likely
	case "50/50":
		return FiftyFifty
	case "Unlikely":
		return Unlikely
	case "Small Chance":
		return SmallChance
	}

	panic(fmt.Sprintf("invalid odds string: %s", s))
}

// resolveOdds
// Input: 1-100 and the chosed odds
// Output: NoYes
func resolveOdds(n int, chosenOdds Odds) NoYes {
	return n >= int(chosenOdds)
}
